Compensate for digest part of the crypto function deprecation

DDvO opened the pull request:
This PR is part of fixing #14628.

Add deprecation compatibility declarations for message digest functions
~MD2, MD4, MD5, MDC2, RIPEMD160, WHIRLPOOL,~ SHA1, SHA224, SHA256, SHA384 and SHA512.
Also add hints to `SHA256_Init.pod` and `CHANGES.md` how to replace `SHA256()` etc.

Yet maybe it's not worth having deprecation compensation declarations for MD2, MD4, MD5, MDC2, RIPEMD160, and WHIRLPOOL since anyway they should not have been used anymore since long. And maybe code like
```
EVP_Digest(d, n, md, NULL, EVP_md5(), NULL)
```
does not even work anymore since the legacy digest have been moved to the legacy provider?

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated

slontis commented:
I agree with the notion of not doing this for any algorithm that sits inside legacyprov.c.

> does not even work anymore since the legacy digest have been moved to the legacy provider?

In relation to EVP_Digest this would not work because
(1) The legacy provider is not loaded
(2) The legacy provider is not in the default NULL library context.. You would need to pass a fetched algorithm into the EVP_Digest to do this.

slontis reviewed (commented):
Not sure that OPENSSL_COMPENSATE_DEPRECATION is a good idea..
Wouldn't we just be better off with helper API's going forward?

slontis commented on doc/man3/MD5.pod:
```diff
@@ -49,7 +49,13 @@ L<openssl_user_macros(7)>:
+can continue to be used by defining B<OPENSSL_COMPENSATE_DEPRECATION>.
+They can also be replaced by, e.g.,
+
+    EVP_Digest(data, count, md_buf, NULL, EVP_md5(), NULL)
```
This should really be a fetch instead of EVP_md5()
EVP_Digest_ex() could pass a libctx, name and propq?

DDvO replied:
Yeah, this is meanwhile taken over by `EVP_Q_digest()`.

DDvO commented:
> I agree with the notion of not doing this for any algorithm that sits inside legacyprov.c.

Fine, so I will remove the compensation decls for MD5 and other legacy digests.

> > does not even work anymore since the legacy digest have been moved to the legacy provider?
> 
> In relation to EVP_Digest this would not work because
> (1) The legacy provider is not loaded
> (2) The legacy provider is not in the default NULL library context.. You would need to pass a fetched algorithm into the EVP_Digest to do this.

This puzzles me since
`EVP_md5(void)` seems still available in `lecacy_md5.c` and
`EVP_Digest()` appears to work without lib contexts, at least it does not need to fetch MD5.

levitte commented:
> I agree with the notion of not doing this for any algorithm that sits inside legacyprov.c.

Does that mean we could deprecate `EVP_md2()`, `EVP_md4()`, `EVP_mdc2()`, `EVP_whirpool()`, `EVP_ripemd160()`, and oh look at what else we've moved to the legacy provider:wink:

paulidale commented on include/openssl/sha.h:
```diff
@@ -82,6 +88,13 @@ OSSL_DEPRECATEDIN_3_0 unsigned char *SHA256(const unsigned char *d, size_t n,
+#   define SHA224(d, n, md) \
+    (EVP_Digest(d, n, md, NULL, EVP_sha224(), NULL) ? md : NULL)
+#   define SHA256(d, n, md) \
+    (EVP_Digest(d, n, md, NULL, EVP_sha256(), NULL) ? md : NULL)
```
`EVP_sha256()` should be a string `"SHA256"` or `SN_sha256` and the algorithm fetched in `EVP_Digest()`.

Similarly everywhere else :grinning:

DDvO replied:
Hmm, `EVP_Digest()` was already there!
Can't we simply keep and use it as-is?
Of course, I can add `EVP_Q_digest()` which takes a string instead, plus a libctx and query string:
```
int EVP_Q_digest(const void *data, size_t count,
                 unsigned char *md, unsigned int *size,
                 OSSL_LIB_CTX *libctx, const char *name, const char *propq);
```

t8m replied:
I think adding the EVP_Q_digest with the parameters as you propose is a good idea.

DDvO replied:
Done.

artem replied:
Sorry for bringing up this thread, but what's the problem with avoiding `EVP_sha256()`? And generally why would `EVP_Q_Digest()` be preferred over `EVP_Digest()`, since the former wraps around the latter?

Asking this from an API user's perspective since I'm not much familiar with ossl internals

DDvO replied:
I'm not an expert on the internals myself, so can just answer from user' perspective.
As far as I understand, the idea behind all those API changes was that algorithm names should no more be hard-coded in the function/macro names, in order to make the API fully crypto agile.
Though I think this would have been possible also with `EVP_Digest()`.

t8m replied:
The EVP_sha256() uses so called implicit fetch of the algorithm. In the context of EVP_Digest vs EVP_Q_Digest it does not mean much apart from the fact, that EVP_Q_Digest can be used with any provided hash even such that OpenSSL libcrypto itself does not know anything about. It just needs to be available within some loaded third party provider.

DDvO commented:
As decided today, I've removed again the compatibility macros for the outdated digest algorithms.
Also have taken he liberty to remove the 'hold: need otc decision' label myself, and rebased this PR.

DDvO commented:
BTW, shouldn't `EVP_Digest()` and `EVP_DigestInit_ex()` be deprecated 
as they (optionally) use an engine parameter?
Looks like `EVP_Q_digest()` and `EVP_DigestInit_ex2()` are reasonable replacements.

t8m commented:
> BTW, shouldn't `EVP_Digest()` and `EVP_DigestInit_ex()` be deprecated
> as they (optionally) use an engine parameter?

There were some discussions around it but AFAIK we did not decide to do that as the engine parameter is just optional. Even if we drop the engine support completely we can keep these functions and opaque ENGINE type or keep them just as macros with this parameter ignored.

DDvO commented:
> > BTW, shouldn't `EVP_Digest()` and `EVP_DigestInit_ex()` be deprecated
> > as they (optionally) use an engine parameter?
> 
> There were some discussions around it but AFAIK we did not decide to do that as the engine parameter is just optional.

I see.
The situation has changed somewhat, though, since now there is a viable replacement of  `EVP_Digest()`, namely `EVP_Q_digest()`. But still it does not hurt to keep the former.

> Even if we drop the engine support completely we can keep these functions and opaque ENGINE type or keep them just as macros with this parameter ignored.

True. And when the engines are gone, a caller will even be unable to provide a non-NULL `ENGINE *` `argument`, so ignoring the parameter will be safe.

DDvO commented:
According to https://github.com/openssl/openssl/issues/14628#issuecomment-819247760, removed all the `OPENSSL_COMPENSATE_DEPRECATION` business also from this PR.

slontis commented on doc/man3/SHA256_Init.pod:
```diff
@@ -49,7 +49,12 @@ L<openssl_user_macros(7)>:
-and L<EVP_DigestFinal_ex(3)>.
+and L<EVP_DigestFinal_ex(3)>, or L<EVP_Q_digest(3)>.
+SHA1(), SHA224(), SHA256(), SHA384(), and SHA256()
+can continue to be used by defining B<OPENSSL_COMPENSATE_DEPRECATION>.
```
OPENSSL_COMPENSATE_DEPRECATION?
Why not just have them always there?

DDvO replied:
Opos, this was I leftover I forgot to remove - done now.
Last week we had decided to go without `OPENSSL_COMPENSATE_DEPRECATION`.

DDvO commented:
Anything more to do on this PR, or can it be approved now?

slontis commented on crypto/sha/sha1_one.c:
```diff
@@ -18,7 +18,7 @@
 #include <openssl/sha.h>
 
-unsigned char *SHA1(const unsigned char *d, size_t n, unsigned char *md)
+unsigned char *SHA1_int(const unsigned char *d, size_t n, unsigned char *md)
```
ossl_sha1() maybe if this is internal

DDvO replied:
Good thought - done.

slontis commented on include/internal/deprecated.h:
```diff
@@ -26,5 +26,8 @@
 # include <openssl/macros.h>
+# include <stddef.h>
+
+unsigned char *SHA1_int(const unsigned char *d, size_t n, unsigned char *md);
```
Dont know that this is the right place for this..

DDvO replied:
Moved to `include/crypto/sha.h`

slontis commented:
@openssl/otc - is everyone happy with this?

DDvO commented:
> @openssl/otc - is everyone happy with this?

@paulidale or @t8m, can you approve?

levitte commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -531,9 +534,10 @@ Returns 1 for success or 0 for failure.
 EVP_DigestUpdate(),
-EVP_DigestFinal_ex()
+EVP_DigestFinal_ex(), and
+EVP_Digest()
```
What about EVP_Q_digest()?

DDvO replied:
Oops, added there.

levitte commented on include/crypto/sha.h:
```diff
@@ -12,10 +12,13 @@
 int sha512_224_init(SHA512_CTX *);
 int sha512_256_init(SHA512_CTX *);
 int ossl_sha1_ctrl(SHA_CTX *ctx, int cmd, int mslen, void *ms);
+# endif
```
Not sure why you find it necessary to deprecate internal functions...

DDvO replied:
Me neither. The CI run `enable_non-default_options` was failing like this:
```
 In file included from crypto/evp/e_des3.c:22:
include/crypto/sha.h:18:21: error: unknown type name 'SHA512_CTX'; did you mean 'SHA512_ASM'?
   18 | int sha512_224_init(SHA512_CTX *);
      |                     ^~~~~~~~~~
      |                     SHA512_ASM
include/crypto/sha.h:19:21: error: unknown type name 'SHA512_CTX'; did you mean 'SHA512_ASM'?
   19 | int sha512_256_init(SHA512_CTX *);
      |                     ^~~~~~~~~~
      |                     SHA512_ASM
include/crypto/sha.h:20:20: error: unknown type name 'SHA_CTX'; did you mean 'SCT_CTX'?
   20 | int ossl_sha1_ctrl(SHA_CTX *ctx, int cmd, int mslen, void *ms);
      |                    ^~~~~~~
      |                    SCT_CTX
```
so I thought this was due to deprecation, but it could also have been due to a missing `#include` which I added meanwhile, so let's see how it works without that guard...

DDvO replied:
Right, the `# ifndef OPENSSL_NO_DEPRECATED_3_0` was not really needed, so removed it.

slontis commented on CHANGES.md:
```diff
@@ -998,15 +998,18 @@ OpenSSL 3.0
    Use of these low level functions has been informally discouraged
    for a long time.  Applications should use the EVP_DigestInit_ex(3),
-   EVP_DigestUpdate(3) and EVP_DigestFinal_ex(3) functions instead.
+   EVP_DigestUpdate(3) and EVP_DigestFinal_ex(3) functions instead or simply,
```
This doesn't read correctly.. Maybe something like

init(), update() and final() functions instead. Alternatively EVP_Q_digest() can be used().
SHA1,SHA224 .... have changed from functions to macros that use EVP_Q_digest().

DDvO replied:
Good point. I've re-worked the 2nd half of the paragraph to:
```
   Alternatively, the quick one-shot function L<EVP_Q_digest(3)> can be used.
   SHA1, SHA224, SHA256, SHA384 and SHA512 have changed from functions to macros
   like this: (EVP_Q_digest(d, n, md, NULL, NULL, "SHA256", NULL) ? md : NULL).
```
which reflects the situation both more succinctly and precisely.

slontis commented on doc/man3/SHA256_Init.pod:
```diff
@@ -49,7 +49,12 @@ L<openssl_user_macros(7)>:
-and L<EVP_DigestFinal_ex(3)>.
+and L<EVP_DigestFinal_ex(3)>, or L<EVP_Q_digest(3)>.
+SHA1(), SHA224(), SHA256(), SHA384(), and SHA256()
+can continue to be used by defining B<OPENSSL_COMPENSATE_DEPRECATION>.
```
This comment is still here?

DDvO replied:
Ops, yes. Re-worked this in the same way as above to:
```
and L<EVP_DigestFinal_ex(3)>, or the quick one-shot function L<EVP_Q_digest(3)>.
SHA1(), SHA224(), SHA256(), SHA384(), and SHA256()
can continue to be used. They can also be replaced by, e.g.,

    (EVP_Q_digest(d, n, md, NULL, NULL, "SHA256", NULL) ? md : NULL)
```
Doing so, I noticed that the synopsis of the following functions had to be moved up because they are no more deprecated. I've also updated their arg names and types:
```
 unsigned char *SHA1(const void *data, size_t count, unsigned char *md_buf);
 unsigned char *SHA224(const void *data, size_t count, unsigned char *md_buf);
 unsigned char *SHA256(const void *data, size_t count, unsigned char *md_buf);
 unsigned char *SHA384(const void *data, size_t count, unsigned char *md_buf);
 unsigned char *SHA512(const void *data, size_t count, unsigned char *md_buf);
```

slontis commented:
Looks ok to me now - I do wonder if we should be namespacing the macros

DDvO commented:
> Looks ok to me now - 

Pleased to hear.

>I do wonder if we should be namespacing the macros

We should not - part of the discussion at #14628 and during the related OTC meeting was that there are user code bases that need to compile cleanly both with 1.1.1 and 3.0 without adapting the code forth and back. To this end the existing names like `SHA256` should not change.

DDvO commented:
Ping @openssl/otc for approval

slontis commented on crypto/evp/digest.c:
```diff
@@ -629,6 +629,20 @@ int EVP_Digest(const void *data, size_t count,
 
+int EVP_Q_digest(const void *data, size_t count,
+                 unsigned char *md, unsigned int *size,
+                 OSSL_LIB_CTX *libctx, const char *name, const char *propq)
```
Can we be consistent at least with the _Q functions in the other PR's and have the libctx and propq in the same place :)

DDvO replied:
> Can we be consistent at least with the _Q functions in the other PR's and have the libctx and propq in the same place :)

The libctx and propq are placed around the digest alg name, as done also for the mac alg name of `EVP_Q_mac()`:
```
unsigned char *EVP_Q_mac(OSSL_LIB_CTX *libctx, const char *name, const char *propq,
                         const char *subalg, const OSSL_PARAM *params, ...
```
The only discrepancy is that for `EVP_Q_digest()` the alg name is at the end.
In this regard I followed the model of `EVP_Digest()`:
```
__owur int EVP_Digest(const void *data, size_t count,
                          unsigned char *md, unsigned int *size,
                          const EVP_MD *type, ENGINE *impl);
```
so IMHO the arg order is a consistent with related functions as it can be.

t8m replied:
IMO there is no reason to follow the EVP_Digest argument order here. I'd make it as much as possible consistent with EVP_Q_mac and disregard the EVP_Digest.

DDvO replied:
Done

slontis commented on doc/man3/SHA256_Init.pod:
```diff
@@ -95,11 +95,12 @@ ANSI X9.30
 =head1 HISTORY
 
-All of these functions were deprecated in OpenSSL 3.0.
+All of these functions except SHA*() were deprecated in OpenSSL 3.0.
```
It should mention which ones were changed to macros here.

DDvO replied:
Done

paulidale commented on CHANGES.md:
```diff
@@ -1070,15 +1070,19 @@ OpenSSL 3.0
+   SHA1, SHA224, SHA256, SHA384 and SHA512 have changed from functions to macros
+   like this: (EVP_Q_digest(NULL, "SHA256", NULL, d, n, md, NULL) ? md : NULL).
 
    *Paul Dale*
```
_ and David von Oheimb_ ?

DDvO replied:
Yes, thanks. Added my name here, and will do also for the changes in #14664.

paulidale reviewed (approved):
Looks good.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Merged - thanks @slontis, @paulidale, et al.

brasic commented:
.
